README.md
# ezSearch
A tool for retrieving product information from multiple platforms at once. Search up to 10 platforms with one click

Current Version: 0.0.1 <br>
Author: Conner Allen <br>
Last Published: 5/8/2022 <br>

# Requirements
The only requirement for this program is <a href="https://nodejs.org/en/">NodeJS</a>

# Installation 
Installation can be done by cloning or downloading this repository

`git clone https://github.com/ConnerAllen-Dev/ezSearch.git`

then installing the needed npm repositories (note these are installed globally, if you want to install them to only the project folder
then nevigate to the repository and remove the `-g` argument in the command below)

`npm install --save -g puppeteer express body-parser figlet inquirer`


# Launch (CLI)
![alt text](https://github.com/ConnerAllen-Dev/ezSearch/blob/main/lib/githubImgs/screenshot1.png?raw=true)

ezSearch has a CLI and GUI available for users to interact with. To use the ezSearch CLI simply `cd` into the folder you cloned or downloaded the repository to then run `node ezSearch`.

# Launch (GUI)
![alt text](https://github.com/ConnerAllen-Dev/ezSearch/blob/main/lib/githubImgs/screenshot2.png?raw=true)

To launch the GUI, simple follow the CLI launch instructions above and once the application has loaded, run `gui`. This will start a express server on "localhost:3000". To view the GUI, simply open your web browser and navigate to "localhost:3000" or "127.0.0.1:3000"

# Use 
Because the GUI is rather self explanitory to navigate, the use section will only apply to the CLI version.

**Commands:**

**"gui"** - Start a express server running on "localhost:3000" that displays ezSearch GUI.

**"search"** - Running search will prompt you for a product name. Simple type in your desired product name and ezSearch will open all windows set to true in the config file.

**"config"** - Simply displays the config file.

**"edit config"** - Will prompt you with "ezedit_>>" which is the ezSeach editor. See below on how to change config settings. <br>
![alt text](https://github.com/ConnerAllen-Dev/ezSearch/blob/main/lib/githubImgs/screenshot3.png?raw=true) <br>
    `SET \<desired website> TO \<desired input>`<br>
    EXAMPLE: `SET google TO true`
    
**"help"** - Displays the help menu.   

**"exit"** - Exits program.


# Features
- Search up to 10 seperate e-commerce platforms with one click
- Automatic VeRo checker, will check the given product name and cross reference the eBay VeRo list. If matched, it will warn you that it is a registered VeRo product
- Customize which platforms you want to search on (Google, Ebay, Amazon are set by default)
- History which keeps trach of your searches and can reseach old searches
- CLI, a command line interface to interacting with program
- GUI, a clean graphical user interface via your web browser at "localhost:3000"

# TODO

- Fix error when user exits webpage before all pages are finished loading (puppeteer navigation error)

# Updates

No updates have been published yet


lib/history.txt
example search one : example search two : exmaple search 3 : schlage connect : capresso infinity conical burr grinder : random : random : random product : black and decker handheld vaccum : 3M : 3M : 3M : Rowenta x-cel steam pocket : random : random product : proctor silex toaster : black and decker rice cooker 3 cup : luxe bidet neo 185 : Milocks keyless entry lock : 24 pc 1oz dropper bottles with funnels : UV sterilizer Box : secura cool touch electric water kettle : dash mini maker griddle : dash mini maker griddle : glamorise 38g bra : mr coffee multi grind blade grinder : random product : random product : random product :  : Rowenta x-cel steam pocket : sunglasses : name : glasses : glasses : mr coffee multi grind blade grinder : random product : sunglasses : sunglasses : sunglasses : sunglasses : sunglasses : sunglasses : sunglasses : sunglasses : sunglasses : sunglasses : test : chefman portable personal fridge

lib/server.js
const os = require('os')
const hd = os.homedir();
const express = require(hd + '\\AppData\\Roaming\\npm\\node_modules\\express');
const bodyParser = require(hd + '\\AppData\\Roaming\\npm\\node_modules\\body-parser');
const mainFunc = require('../ezsearch_gui.js')

const app = express()
app.use(bodyParser.urlencoded({ extended: false }));
app.use(bodyParser.json())
const port = 3000

app.get('/', (req, res) => {
  res.sendFile(__dirname + '/html/index.html')
})

app.get('/config', (req, res) => {
  res.sendFile(__dirname + '/config.txt')
})

app.get('/history', (req, res) => {
  res.sendFile(__dirname + '/history.txt')
})

app.get('/veroList', (req, res) => {
  res.sendFile(__dirname + '/veroList.txt')
})

app.get('/juice.ttf', (req, res) => {
  res.sendFile(__dirname + '/html/Juice Avocado.ttf')
})


//Handlers
app.post('/search', (req, res) => {
	let p = req.body.data;
	console.log(p)
	mainFunc(p);
})



app.listen(port, () => {
  console.log(`\nExpress server started!\nYou can now view the ezsearch GUI via "http://localhost:3000" in your web browser`)
})







ezsearch_gui.js

//Pretty much the same as "main.js" expect sturctered a bit differently to work with the browser GUI


const os = require('os')
const hd = os.homedir();
const puppeteer = require(hd + '\\AppData\\Roaming\\npm\\node_modules\\puppeteer');
const veroChecker = require('./lib/veroCheck_gui.js');
const help = require('./lib/help.js');
const fs = require('fs');

const sleep = ms => {return new Promise(res => {setTimeout(res, ms)})};
//Check config file
let config = JSON.parse(fs.readFileSync('./lib/config.txt').toString());
const mainFunc = async (arg) => {


	//Accepted commands 
	const acceptedCommands = ['search', 'help', 'history', 'config', 'edit config', 'exit'];


	const run = async (product) => {

		let command = arg.split(' : ')[0];
		let data = arg.split(' : ')[1];

		if(command == 'SEARCH'){

			let h = fs.readFileSync('./lib/history.txt').toString();
			let newText = `${h} : ${data}`;
			fs.writeFileSync('./lib/history.txt', newText);
			let x = data;

			//Vero checker
			veroChecker(data);

			//Make puppeteer search
			const browser = await puppeteer.launch({headless: false,  args: ['--start-maximized']});
			
			config.urls.forEach(async (item, i) => {
				if(config.websites[Object.keys(config.websites)[i]] == "true"){
					let url = item.replace('${res.main}', data);
					const page6 = await browser.newPage();
					await page6.setViewport({width:0, height:0});
					await sleep(1000)
					await page6.goto(url);
				}
			})

		}
	
	}
	run()
}


module.exports = mainFunc;

lib/serverStart.js
const os = require('os')
const hd = os.homedir();
const express = require(hd + '\\AppData\\Roaming\\npm\\node_modules\\express');
const bodyParser = require(hd + '\\AppData\\Roaming\\npm\\node_modules\\body-parser');
const mainFunc = require('../ezsearch_gui.js')

const app = express()
app.use(bodyParser.urlencoded({ extended: false }));
app.use(bodyParser.json())
const port = 3000

app.get('/', (req, res) => {
  res.sendFile(__dirname + '/html/index.html')
})

app.get('/config', (req, res) => {
  res.sendFile(__dirname + '/config.txt')
})

app.get('/history', (req, res) => {
  res.sendFile(__dirname + '/history.txt')
})

app.get('/veroList', (req, res) => {
  res.sendFile(__dirname + '/veroList.txt')
})

app.get('/juice.ttf', (req, res) => {
  res.sendFile(__dirname + '/html/Juice Avocado.ttf')
})


//Handlers
app.post('/search', (req, res) => {
	let p = req.body.data;
	console.log(p)
	mainFunc(p);
})



const startSever = () => {
	app.listen(port, () => {
	  console.log(`\nExpress server started!\nYou can now view the ezsearch GUI via "http://localhost:3000" in your web browser`)
	})
}

module.exports = startSever;








ezsearch.js
const os = require('os')
const hd = os.homedir();
const inquirer = require(hd + '\\AppData\\Roaming\\npm\\node_modules\\inquirer');
const puppeteer = require(hd + '\\AppData\\Roaming\\npm\\node_modules\\puppeteer');
const figlet = require(hd + '\\AppData\\Roaming\\npm\\node_modules\\figlet');
const veroChecker = require('./lib/veroCheck.js');
const help = require('./lib/help.js');
const startServer = require('./lib/serverStart.js');
const fs = require('fs');

const sleep = ms => {return new Promise(res => {setTimeout(res, ms)})}
const mainFunc = async () => {
	figlet('\\---ezSearch---/', function(err, data) {
	    if (err) {
	        console.log('Something went wrong...');
	        console.dir(err);
	        return;
	    }
	    console.log(data);
	    console.log('Author: Conner Allen');
	    console.log('Version: 0.0.1');
	    console.log('Welcome to the ezSearch CLI. Run the command "help" if you have any questions about use.');
	    console.log('If you would prefer a GUI, then you can run "gui" to begin an instance in your browser. Otherwise enjoy!\n');

	});

	await sleep(3000);

	//Inquirer questions
	const q = {
		name: 'main',
		type: 'input',
		message: 'ezsearch_>>'
	}
	const q2 = {
		name: 'main',
		type: 'input',
		message: 'Select a product by number: '
	}
	const q3 = {
		name: 'main',
		type: 'input',
		message: 'Product to search: '
	}
	const q4 = {
		name: 'main',
		type: 'input',
		message: 'ezedit_>> '
	}

	//Accepted commands 
	const acceptedCommands = ['search', 'help', 'history', 'config', 'edit config', 'gui','exit'];

	//Check config file
	let config = JSON.parse(fs.readFileSync('./lib/config.txt').toString());

	const run = (product) => {
		inquirer.prompt([q])
		.then(async res => {
			//Check for valid command
			if(!acceptedCommands.includes(res.main)){
				console.log(`!!!!!!! WARNING !!!!!!! - ERROR : Command not recognized. Run help to see available commands`);
				run();
				return;
			}

			//History search
			if(res.main == 'history'){
				//Add search to history
				let existingText = fs.readFileSync('./lib/history.txt').toString();

				let arr = existingText.split(':')

				arr.forEach((item, i) => {
					console.log(`[${i}]   :::::::::   ${item}`)
				})

				inquirer.prompt([q2])
				.then(async res => {
					//Check for additional argument
					let x;
					if(res.main.includes('AND')){
						x = `${arr[res.main]} ${res.main.split('AND')[1]}` 
					}else{
						x = arr[res.main];
					}
					//Make puppeteer search
					const browser = await puppeteer.launch({headless: false,  args: ['--start-maximized']});
					
					config.urls.forEach(async (item, i) => {
						if(config.websites[Object.keys(config.websites)[i]] == "true"){
							let url = item.replace('${res.main}', res.main);
							const page = await browser.newPage();
							await page.setViewport({width:0, height:0});
							await page.setViewport({ width: 1920, height: 1080});
							await sleep(1000)
							await page.goto(url);
						}
					})

					await browser.on('disconnected', () => {
						console.log('Disconnected...')
						run();
					})

				})
				.catch(err => {console.log(err)})
			}
			if(res.main == 'search'){
				inquirer.prompt([q3])
				.then(async res => {
					let h = fs.readFileSync('./lib/history.txt').toString();
					let newText = `${h} : ${res.main}`;
					fs.writeFileSync('./lib/history.txt', newText);
					let x = res.main;

					//Vero checker
					veroChecker(res.main);

					//Make puppeteer search
					const browser = await puppeteer.launch({headless: false,  args: ['--start-maximized']});

					config.urls.forEach(async (item, i) => {
						if(config.websites[Object.keys(config.websites)[i]] == "true"){
							let url = item.replace('${res.main}', res.main);
							const page6 = await browser.newPage();
							await page6.setViewport({width:0, height:0});
							await sleep(1000)
							await page6.goto(url);
						}
					})

					await browser.on('disconnected', () => {
						console.log('Disconnected...');
						run();
					})

				})
				.catch(err => {console.log(err)})
			}

			//Bring up help menu
			if(res.main == 'help'){
				help();
				run();
			}

			//Bring up config menu
			if(res.main == 'config'){
				let d = JSON.parse(fs.readFileSync('./lib/config.txt').toString());
				console.log(d)
				run();
			}
				
			//Edit config file
			if(res.main == 'edit config'){
				inquirer.prompt([q4])
				.then(res => {
					let site = res.main.split('SET ')[1].split(' TO ')[0];
					config.websites[site] = res.main.split('SET ')[1].split(' TO ')[1];
					fs.writeFileSync('./lib/config.txt', JSON.stringify(config));
					console.log('Changes saved!')
					run();
				})
				.catch(err => console.log(err));
			}

			if(res.main =='gui'){
				console.log('Loading ezsearch GUI...');
				startServer();
			}

			//Exit 
			if(res.main == 'exit'){
				console.log('Bye!')
			}
			
			
		})
		.catch(err => console.log(err))
			
	}
	run()
}

mainFunc();

